crates/matrix-sdk-common/src/deserialized_responses.rs

```diff
@@ -666,15 +669,26 @@ pub struct UnableToDecryptInfo {
     pub session_id: Option<String>,
 
     /// Reason code for the decryption failure
-    #[serde(default = "unknown_utd_reason")]
+    #[serde(default = "unknown_utd_reason", deserialize_with = "ok_reason_or_default")]
     pub reason: UnableToDecryptReason,
 }
 
 fn unknown_utd_reason() -> UnableToDecryptReason {
     UnableToDecryptReason::Unknown
 }
 
-/// Reason code for a decryption failure
+/// Provides basic backward compatibility for deserializing older serialized
+/// `UnableToDecryptReason` values. If the serialized format has changed in a
+/// way that causes deserialization to fail, the reason will default to
+/// `Unknown`.
+fn ok_reason_or_default<'de, D>(d: D) -> Result<UnableToDecryptReason, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let v: serde_json::Value = Deserialize::deserialize(d)?;
+    Ok(UnableToDecryptReason::deserialize(v).unwrap_or(UnableToDecryptReason::Unknown))
+}
```

Comment:
This feels like the wrong approach for three reasons:

 1. It only applies to `UnableToDecryptReason`s that are embedded in `UnableToDecryptInfo`, so this still breaks backwards compatibility for `UnableToDecryptReason`s that have been serialized in any other environment.
 2. It turns existing `MissingMegolmSession`s into `UnableToDecryptReason::Unknown`, which seems like a disappointing loss of data.
 3. It's a bit of a sledgehammer: turning every single decryption error into a `UnableToDecryptReason::Unknown` seems like another loss of data.

I think points 2 and 3 can be fixed by manipulating the json `Value` before calling `UnableToDecryptReason::deserialize(v)`, but I don't know how to fix point 1 without either (a) creating and maintaining another copy of `UnableToDecryptReason` (so that one version can have a custom deserializer that fiddles with the serialized format and then uses the default deserializer of the second version) or (b) hand-writing an entire deserializer. Will ask in the rust room.

Reply:
Having discussed this internally: it seems unlikely that anyone will be serializing `UnableToDecryptReason` outside `UnableToDecryptInfo`, and in any case we aren't yet doing stable releases, so we can disregard point 1.

I'd still quite like us to address points 2 & 3. Something like:


```suggestion
pub fn deserialize_utd_reason<'de, D>(d: D) -> Result<UnableToDecryptReason, D::Error>
where
    D: serde::Deserializer<'de>,
{
    // Start by deserializing as to an untyped JSON value.
    let v: serde_json::Value = Deserialize::deserialize(d)?;

    // Backwards compatibility: `MissingMegolmSession` used to be stored without the withheld code.
    if v.as_str().is_some_and(|s| s == "MissingMegolmSession") {
        return Ok(UnableToDecryptReason::MissingMegolmSession(None));
    }

    // Otherwise, use the derived deserialize impl to turn the JSON into a UnableToDecryptReason
    serde_json::from_value::<UnableToDecryptReason>(v).map_err(serde::de::Error::custom)
}
```

(note also, improved name and the fact that it should probably be `pub`.)